plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
+                DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucketName).withKeys(objectKeysBatch);
+                if (requesterPaysEnabled) {
+                    // TODO use deleteObjectRequest.setRequesterPays() when https://github.com/aws/aws-sdk-java/issues/1219 is fixed
+                    // currently the method exists, but is ineffective (doesn't set the required HTTP header)
+                    deleteObjectsRequest.putCustomRequestHeader(REQUESTER_PAYS_HEADER, REQUESTER_PAYS);
+                }
+                s3.deleteObjects(deleteObjectsRequest);
+            }
+            catch (MultiObjectDeleteException e) {
+                e.getErrors().stream().map(MultiObjectDeleteException.DeleteError::getKey).forEach(objectKeysFailedToDelete::add);
+            }
         }
 
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        // Clean up eventual objects which the deletion has failed while doing the batched deletion
+        while (!objectKeysFailedToDelete.isEmpty()) {
+            Set<String> objectKeysStillPendingForDeletion = new HashSet<>();
+            for (List<String> objectKeysBatch : Iterables.partition(objectKeysFailedToDelete, BATCH_DELETE_OBJECTS_REQEST_MAX_PAGE_SIZE)) {
+                try {
+                    DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucketName).withKeys(objectKeysBatch.toArray(String[]::new));
+                    if (requesterPaysEnabled) {
+                        // TODO use deleteObjectRequest.setRequesterPays() when https://github.com/aws/aws-sdk-java/issues/1219 is fixed
+                        // currently the method exists, but is ineffective (doesn't set the required HTTP header)
+                        deleteObjectsRequest.putCustomRequestHeader(REQUESTER_PAYS_HEADER, REQUESTER_PAYS);
+                    }
+                    s3.deleteObjects(deleteObjectsRequest);
+                }
+                catch (MultiObjectDeleteException e) {
+                    e.getErrors().stream().map(MultiObjectDeleteException.DeleteError::getKey).forEach(objectKeysStillPendingForDeletion::add);
+                }
+            }
+
+            // In the unlikely scenario where deleteObjects call cannot make progress, avoid infinite loop
+            if (objectKeysFailedToDelete.size() == objectKeysStillPendingForDeletion.size()) {
+                throw new RuntimeException("Cannot make progress deleting objects under the path: " + prefix);
+            }
+
+            if (!objectKeysStillPendingForDeletion.isEmpty()) {
+                objectKeysFailedToDelete = new HashSet<>(objectKeysStillPendingForDeletion);
+            }
         }
-        deleteObject(keyFromPath(path) + DIRECTORY_SUFFIX);
 
         return true;
     }
 
+    @VisibleForTesting
+    static Iterator<String> keysFromListing(ListObjectsV2Result listing)
+    {
+        List<String> prefixes = listing.getCommonPrefixes();
```

Comment:
Returning `prefixes` as "keys" in `keysFromListing` seems wrong, because they aren't actually S3 objects. We are expecting to have done a "no path separator" listing which should never include prefix entries in the response. Maybe we should refactor away this code or fail loudly if prefixes are present?